src/ast/passes/probe_analyser.cpp

```diff
@@ -0,0 +1,57 @@
+#include "probe_analyser.h"
+
+namespace bpftrace::ast {
+
+void ProbeAnalyser::visit(Probe &probe)
+{
+  // If the probe has a single multi-expanded kprobe attach point, check if
+  // there's another probe with a single multi-expanded kretprobe attach point
+  // with the same target. If so, use session expansion if available.
+  // Also, we do not allow predicates in any of the probes for now.
+  if (bpftrace_.feature_->has_kprobe_session() &&
+      probe.attach_points.size() == 1 &&
+      probetype(probe.attach_points[0]->provider) == ProbeType::kprobe &&
+      probe.pred == nullptr) {
+    // Session probes use the same attach mechanism as multi probes so the
+    // attach point must be multi-expanded.
+    auto &ap = *probe.attach_points[0];
+    if (ap.expansion != ExpansionType::MULTI)
+      return;
+
+    for (Probe *other_probe : ctx_.root->probes) {
+      // Other probe must also have a single multi-expanded attach point and no
+      // predicate
+      if (other_probe->attach_points.size() != 1 || other_probe->pred)
+        continue;
+      auto &other_ap = *other_probe->attach_points[0];
+      if (probetype(other_ap.provider) != ProbeType::kretprobe ||
+          other_ap.expansion != ExpansionType::MULTI) {
+        continue;
+      }
+
+      if (ap.target == other_ap.target && ap.func == other_ap.func) {
+        ap.expansion = ExpansionType::SESSION;
+        ap.ret_probe = other_probe;
```

Comment:
I really think that it would be better to do an actual transformation here (e.g. remove the probe, replace with another AST node that reflects a `SessionProbe` or other such specialized case) rather than leaving breadcrumbs and an implicit requirements that codegen do the right thing with respect to new and old probes.

I know that the current structure doesn't actually let us do this (and I need to finish and submit my own contribution there), but I'm just trying to raise awareness.

Reply:
Good point. We used to have a `Mutator` class but it was removed in #3347 as it was kind of broken.

I think we would be able to identify quite a number of cases where we could do actual AST transformations instead of having classes with a lot of fields specific to certain use-cases only (like `Probe` or so).

Reply:
The new mutation infra is in now. Is it worth trying to use in this PR? Agreed a few other places could use it as well. But in general I'd agree it's better to lower the AST into structurally unambiguous representations when possible.

Reply:
Might be nice but I don't see that as being a blocker - could be done as part of a larger, follow-up refactor.

Reply:
@amscanne @danobi are you ok with merging this in the current form?